selfdrive/car/hyundai/values.py

```diff
@@ -348,35 +344,66 @@ class Buttons:
+  # Returns unique, platform-specific identification codes for a set of versions
+  codes = set()  # (code-Optional[part], date)
   for fw in fw_versions:
-    code_match, date_match = (PLATFORM_CODE_FW_PATTERN.search(fw),
-                              DATE_FW_PATTERN.search(fw))
+    code_match, part_match, date_match = (PLATFORM_CODE_FW_PATTERN.search(fw),
+                                          PART_NUMBER_FW_PATTERN.search(fw),
+                                          DATE_FW_PATTERN.search(fw))
     if code_match is not None:
-      code = code_match.group()
+      code: bytes = code_match.group()
+      part = part_match.group() if part_match else None
       date = date_match.group() if date_match else None
-      codes[code].add(date)
+      if part is not None:
+        # part number starts with generic ECU part type, add what is specific to platform
+        code += b"-" + part[-5:]
 
-  # Create platform codes for all dates inclusive if ECU has FW dates
-  final_codes = set()
-  for code, dates in codes.items():
-    # Radar and some cameras don't have FW dates
-    if None in dates:
-      final_codes.add(code)
-      continue
+      codes.add((code, date))
+  return codes
 
-    try:
-      parsed = {datetime.strptime(cast(bytes, date).decode()[:4], '%y%m') for date in dates}
-    except ValueError:
-      cloudlog.exception(f'Error parsing date in FW versions: {code!r}, {dates}')
-      final_codes.add(code)
-      continue
 
-    for monthly in rrule.rrule(rrule.MONTHLY, dtstart=min(parsed), until=max(parsed)):
-      final_codes.add(code + b'-' + monthly.strftime('%y%m').encode())
+def match_fw_to_car_fuzzy(fw_versions_dict, log=True) -> Set[str]:
+  # Non-electric CAN FD platforms often do not have platform code specifiers needed
+  # to distinguish between hybrid and ICE. All EVs so far are either exclusively
+  # electric or specify electric in the platform code.
+  invalid = set(CANFD_CAR - EV_CAR)
+  candidates = FW_VERSIONS
 
-  return final_codes
+  for candidate, fws in candidates.items():
+    for ecu, expected_versions in fws.items():
+      addr = ecu[1:]
+      # Only check ECUs expected to have platform codes
+      if ecu[0] not in PLATFORM_CODE_ECUS:
+        continue
+
+      # Expected platform codes & dates
+      codes = get_platform_codes(expected_versions)
+      expected_platform_codes = {code for code, _ in codes}
+      expected_dates = {date for _, date in codes if date is not None}
+
+      # Found platform codes & dates
+      codes = get_platform_codes(fw_versions_dict.get(addr, set()))
+      found_platform_codes = {code for code, _ in codes}
+      found_dates = {date for _, date in codes if date is not None}
+
+      # Check platform code + part number matches for any found versions
+      if not any(found_platform_code in expected_platform_codes for found_platform_code in found_platform_codes):
+        invalid.add(candidate)
+        break
+
+      # Don't check dates if none in database
```

Comment:
which ones don't have dates?

Reply:
I have a test that checks all the FW for each ECU either exists or not, but I kinda wanted to make a list of platforms like for the excluded platforms, I'll change the test

Reply:
the debug script mentioned above should make this obvious

Reply:
added script and test